src/main/scala/io/iohk/ethereum/jsonrpc/server/http/RateLimit.scala

```diff
@@ -1,25 +1,42 @@
 package io.iohk.ethereum.jsonrpc.server.http
 
-import java.time.Clock
-
-import akka.http.scaladsl.model.RemoteAddress
-import com.twitter.util.LruMap
-import io.iohk.ethereum.jsonrpc.server.http.JsonRpcHttpServer.JsonRpcHttpServerConfig
-
-trait RateLimit {
-
-  val config: JsonRpcHttpServerConfig
-
-  val latestRequestTimestamps = new LruMap[RemoteAddress, Long](config.rateLimit.latestTimestampCacheSize)
-
-  val clock: Clock = Clock.systemUTC()
-
-  def isBelowRateLimit(clientAddress: RemoteAddress): Boolean = {
-    val timeMillis = clock.instant().toEpochMilli
-    val latestRequestTimestamp = latestRequestTimestamps.getOrElse(clientAddress, 0L)
+import akka.http.scaladsl.model.{RemoteAddress, StatusCodes}
+import akka.http.scaladsl.server.{Directive0, Route}
+import io.iohk.ethereum.db.cache.SimpleLRU
+import io.iohk.ethereum.jsonrpc.server.http.JsonRpcHttpServer.RateLimitConfig
+import akka.http.scaladsl.server.Directives._
+import io.iohk.ethereum.jsonrpc.JsonRpcError
+import de.heikoseeberger.akkahttpjson4s.Json4sSupport
+import io.iohk.ethereum.jsonrpc.serialization.JsonSerializers
+import org.json4s.{DefaultFormats, Formats, Serialization, native}
+
+
+class RateLimit(config: RateLimitConfig) extends Directive0 with Json4sSupport {
+
+  private implicit val serialization: Serialization = native.Serialization
+  private implicit val formats: Formats = DefaultFormats + JsonSerializers.RpcErrorJsonSerializer
+
+  // It determines whether a request needs to be blocked
+  // Such algebras prevent if-elseif-else boilerplate in the JsonRPCServer code
+  val blockingAlgebra: (RemoteAddress => Boolean) = {
+    if (config.enabled) {
+      val lru = new SimpleLRU[RemoteAddress](
```

Comment:
Can this `val lru` be declared and initialized in the outer `RateLimit` class body? I think that would make it clear that this is the same cache instance always.

Reply:
Sure, but the trick was to *not* create any LRUs in case when ratelimit is disabled in config.

Reply:
But I've got an idea, thanks! Probably we don't even have to extract any IP when the feature is disabled...